crates/turbopack-ecmascript/src/references/unreachable.rs

```diff
@@ -29,20 +45,199 @@ impl CodeGenerateable for Unreachable {
+                            }
+                        }
+                    }
+                    let (parent, [last]) = parent.split_at(parent.len() - 1) else {
+                        unreachable!();
+                    };
+                    if let &AstParentKind::BlockStmt(BlockStmtField::Stmts(start_index)) = last {
+                        [
+                            create_visitor!(exact parent, visit_mut_block_stmt(block: &mut BlockStmt) {
+                                replace(&mut block.stmts, start_index);
+                            }),
+                        ]
+                        .into()
+                    } else if let &AstParentKind::SwitchCase(SwitchCaseField::Cons(start_index)) =
+                        last
+                    {
+                        [
+                            create_visitor!(exact parent, visit_mut_switch_case(case: &mut SwitchCase) {
+                                replace(&mut case.cons, start_index);
+                            }),
+                        ]
+                        .into()
+                    } else {
+                        Vec::new()
+                    }
+                } else {
+                    Vec::new()
+                }
+            }
+        };
 
         Ok(CodeGeneration { visitors }.cell())
     }
 }
+
+struct ExtractDeclarations<'a> {
+    stmts: &'a mut Vec<Stmt>,
+}
+
+impl<'a> VisitMut for ExtractDeclarations<'a> {
+    fn visit_mut_var_decl(&mut self, decl: &mut VarDecl) {
+        let VarDecl {
+            span,
+            kind,
+            declare,
+            decls,
+        } = decl;
+        let mut idents = Vec::new();
+        for decl in take(decls) {
+            collect_idents(&decl.name, &mut idents);
+        }
+        let decls = idents
+            .into_iter()
+            .map(|ident| VarDeclarator {
+                span: ident.span,
+                name: Pat::Ident(BindingIdent {
+                    id: ident,
+                    type_ann: None,
+                }),
+                init: if matches!(kind, VarDeclKind::Const) {
+                    Some(quote!("undefined" as Box<Expr>))
+                } else {
+                    None
+                },
+                definite: false,
+            })
+            .collect();
+        self.stmts.push(Stmt::Decl(Decl::Var(Box::new(VarDecl {
+            span: *span,
+            kind: *kind,
+            declare: *declare,
+            decls,
+        }))));
+    }
+
+    fn visit_mut_fn_decl(&mut self, decl: &mut FnDecl) {
+        let FnDecl {
+            declare,
+            ident,
+            function,
+        } = decl;
+        self.stmts.push(Stmt::Decl(Decl::Fn(FnDecl {
+            declare: *declare,
+            ident: ident.take(),
+            function: function.take(),
+        })));
+    }
+
+    fn visit_mut_fn_expr(&mut self, _: &mut FnExpr) {
+        // Do not walk into function expressions
```

Comment:
We may need to handle `Constructor`, `GetterProp`, `SetterProp`, `MethodProp`, `ClassMethod`, and `PrivateMethod`